File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = deque([])
        n, m = len(grid), len(grid[0])

        def inbound(i, j):
            return i >=0 and j >=0 and i < n and j < m

        vis = [[False for _ in range(m)] for _ in range(n)]

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2:
                    q.append((i, j, 0))
                    vis[i][j] = True

        mins = 0
        while q:
            siz = len(q)
            for _ in range(siz):
                r, c, t = q.popleft()
                mins = max(mins, t)
                for dx, dy in [(0,1), (1,0), (-1,0), (0,-1)]:
                    row, col = r + dx, c + dy
                    if inbound(row, col) and grid[row][col] in (1, 2)\
                     and not vis[row][col]:
                        q.append((row, col, t + 1))
                        grid[row][col] = 2
                        vis[row][col] = True


        if not all(element in (0, 2) for l in grid for element in l):
            return -1
        
        return mins 



        

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public static final int[][] DIRS = {{0,1}, {1,0}, {0,-1}, {-1,0}};
    public int orangesRotting(int[][] grid) {
        // int [][] g = grid.clone();
        // Do this in interview
        Set<String> vis = new HashSet<>();

        Queue<int[]> q = new LinkedList<>();

        int n = grid.length, m = grid[0].length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 2) {
                    vis.add(i + "," + j);
                    q.add(new int[]{i, j, 0});
                }
            }
        }

        int time = 0;

        while (!q.isEmpty()) {
            int [] info = q.poll();
            int r = info[0];
            int c = info[1];
            int t = info[2];

            time = Math.max(time, t);

            for (int [] dir: DIRS) {
                int nr = r + dir[0];
                int nc = c + dir[1];

                if (nr >= 0 && nr < n && nc >= 0 && nc < m && grid[nr][nc] == 1 && !vis.contains(nr + "," + nc))  {
                    q.add(new int [] {nr, nc, t + 1});
                    vis.add(nr + "," + nc);
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    if (!vis.contains(i + "," + j)) return -1;
                }
            }
        }

        return time;
    }
}